### PR TITLE
Add CLI options and runtime support for selection of output snapshot version

### PR DESCRIPTION
**Pull request by svenski123**
This PR adds an optional `--snapshot-version` command line argument to `solana-validator` and `solana-ledger-tool create-snapshot` to allow configuration of the snapshot format version used in generated snapshots.

### CODE REVIEW

**Comment by ryoqun** [comment]
Code looks nice!

Could you add some simple two or three commands to test `--snapshot-version=1.1.0` as well?: https://github.com/solana-labs/solana/blob/075dc078daa8273179eb450939aa9475f49047c7/ci/run-sanity.sh#L29.

I want to do but couldn't so. But this isn't true now! Thanks on working on this!

File: ledger-tool/src/main.rs [hunk 1]
```diff
@@ -653,7 +655,17 @@ fn main() {
...
+    let snapshot_version_arg = Arg::with_name("snapshot_version")
+        .long("snapshot-version")
+        .value_name("SNAPSHOT_VERSION")
+        .validator(|arg| {
+            arg.parse::<SnapshotVersion>()
+                .map(|_| ())
+                .map_err(|e| e.to_string())
+        })
+        .takes_value(true)
+        .default_value(SnapshotVersion::default().into())
```
Comment (ryoqun) [inline ledger-tool/src/main.rs:667]: Heh, nice code deduplication technique. :)

File: ledger/src/snapshot_utils.rs [hunk 2]
```diff
@@ -37,14 +37,20 @@ pub const TAR_VERSION_FILE: &str = "version";
 const MAX_SNAPSHOT_DATA_FILE_SIZE: u64 = 32 * 1024 * 1024 * 1024; // 32 GiB
 const VERSION_STRING_V1_1_0: &str = "1.1.0";
 const VERSION_STRING_V1_2_0: &str = "1.2.0";
-const OUTPUT_SNAPSHOT_VERSION: SnapshotVersion = SnapshotVersion::V1_2_0;
+const DEFAULT_SNAPSHOT_VERSION: SnapshotVersion = SnapshotVersion::V1_2_0;
 
 #[derive(Copy, Clone, Eq, PartialEq, Debug)]
 pub enum SnapshotVersion {
```
Comment (ryoqun) [inline ledger/src/snapshot_utils.rs:43]: Because this is now exposed to the cli, could you make this more user friendly? `impl FromStr for SnapshotVersion`:
https://github.com/solana-labs/solana/pull/10536/files#diff-0f0af8fcfa1444e1b07d671c5bfa4a69R70

- `fmt!` to embed actual bad version string user supplied: "unsupported snapshot version"
- Allow `v` prefixing like `v1.2.0`. Precedent: https://github.com/solana-labs/solana/pull/8297

Comment (svenski123) [reply ledger/src/snapshot_utils.rs:43]: Caller to str::parse supplies the input string and thus has it, no need for Error object to allocate heap memory & copy input.

For CLI, input validator constructs user friendly error message flagging supplied string for user.

FromStr now strips off /^[vV]/; note this necessarily also applies to the contents of the version file in the snapshot archive. Snapshot versions will not change with every minor release, having them be so similar is possibly misleading.

In any event the purpose of this PR is not to define the snapshot version identifier but merely to expose it (whatever it is) to the command line for use as an optional advanced / debug option.

**Comment by svenski123** [comment]
As for run-sanity.sh, I would rather not make any changes to the CI system in this PR for a number of reasons.

I have changed the bank_forks tests to cover both snapshot versions which should ensure adequate testing coverage.

File: clap-utils/src/input_validators.rs [hunk 3]
```diff
@@ -6,50 +6,86 @@ use solana_sdk::{
     pubkey::Pubkey,
     signature::{read_keypair_file, Signature},
 };
+use std::fmt::Display;
 use std::str::FromStr;
 
+fn is_parsable_generic<U, T>(string: T) -> Result<(), String>
```
Comment (ryoqun) [inline clap-utils/src/input_validators.rs:12]: @svenski123 Nice! Could you split off this generic `clap-utils` parsing message improvement as a separate pr? I think this is useful for v1.1 as well and we want to manage logically-separatable changes as separate PRs. Like #10096, I'll land that pr first.

Comment (svenski123) [reply clap-utils/src/input_validators.rs:12]: Done in #10599, will rebase this once that's been merged.

File: core/tests/bank_forks.rs [hunk 4]
```diff
@@ -38,400 +39,455 @@ mod tests {
...
-            genesis_config_info,
-        }
-    }
+    macro_rules! DEFINE_SNAPSHOT_VERSION_PARAMETERIZED_TEST_FUNCTIONS {
+        ($x:ident) => {
+            #[allow(non_snake_case)]
+            mod $x {
+                use super::*;
+
+                const SNAPSHOT_VERSION: SnapshotVersion = SnapshotVersion::$x;
```
Comment (ryoqun) [inline core/tests/bank_forks.rs:48]: @svenski123 Could you try to minimize the code put under `macro_rules!`?  I know this is the straight-forward way to do these tests with the unfortunate lack of property-based testing with cargo test facility. So, I understand your intention. :)

But, I think minimizing the macro code is doable and desirable for following reasons:

- Reduces diff (This is important metrics, considering possible back-ports)
- Generally macro code is hard to develop because of indirect compilation messages.

My suggestion is like this if you want to keep nice test result output(https://buildkite.com/solana-labs/solana/builds/26005#41e8769e-a92c-47b3-a621-e2969d08d70e/76-1810):

```rust
-    #[test]
-    fn test_bank_forks_snapshot_n() {
+    fn test_bank_forks_snapshot_n_inner(version) {
     }
     ...
     macro_rules! DEFINE_SNAPSHOT_VERSION_PARAMETERIZED_TEST_FUNCTIONS { 
         ($x:ident) => { 
             #[allow(non_snake_case)] 
             mod $x { 
                 use super::*; 
                 const SNAPSHOT_VERSION: SnapshotVersion = SnapshotVersion::$x;

                 #[test]
                 fn test_bank_forks_snapshot_n() {
                     test_bank_forks_snapshot_n_inner(SNAPSHOT_VERSION)
                 }
...
```

Or I'm fine even with this:

```rust
-#[test]
 fn test_bank_forks_snapshot_n_inner(version) {
 }

+#[test]
 fn test_bank_forks_snapshot_n() {
    test_bank_forks_snapshot_n_inner(V1_1_0)
    test_bank_forks_snapshot_n_inner(V1_2_0)
 }
...
```

Comment (svenski123) [reply core/tests/bank_forks.rs]: Refactored with only the stub functions in the macro

File: ledger-tool/src/main.rs [hunk 5]
```diff
@@ -1094,7 +1111,7 @@ fn main() {
                     });
 
                     let storages: Vec<_> = bank.get_snapshot_storages();
-                    snapshot_utils::add_snapshot(&temp_dir, &bank, &storages)
+                    snapshot_utils::add_snapshot(&temp_dir, &bank, &storages, snapshot_version)
```
Comment (ryoqun) [inline ledger-tool/src/main.rs:1114]: nits: how about printing snapshot_version too here?: https://github.com/solana-labs/solana/pull/10536/files#diff-3125b196bc2e46166465885cd9ad66adR1102

File: core/tests/bank_forks.rs [hunk 6]
```diff
@@ -118,18 +159,27 @@ mod tests {
...
         solana_logger::setup();
         // Set up snapshotting config
-        let mut snapshot_test_config = setup_snapshot_test(1);
+        let mut snapshot_test_config = SnapshotTestConfig::new(snapshot_version, 1);
 
         let bank_forks = &mut snapshot_test_config.bank_forks;
         let accounts_dir = &snapshot_test_config.accounts_dir;
         let snapshot_config = &snapshot_test_config.snapshot_config;
         let mint_keypair = &snapshot_test_config.genesis_config_info.mint_keypair;
+        let snapshot_version = bank_forks
```
Comment (ryoqun) [inline core/tests/bank_forks.rs:178]: nits: Why is this reassignment of `snapshot_version` needed? We can use just `snapshot_version` in argument at here: https://github.com/solana-labs/solana/pull/10536/files#diff-60ffd8c11b11dddb54a2816436a96d70R212, or not?

File: core/tests/bank_forks.rs [hunk 7]
```diff
@@ -206,24 +257,34 @@ mod tests {
...
-        let mut snapshot_test_config = setup_snapshot_test(1);
+        let mut snapshot_test_config = SnapshotTestConfig::new(snapshot_version, 1);
 
         let bank_forks = &mut snapshot_test_config.bank_forks;
         let accounts_dir = &snapshot_test_config.accounts_dir;
         let snapshots_dir = &snapshot_test_config.snapshot_dir;
         let snapshot_config = &snapshot_test_config.snapshot_config;
         let mint_keypair = &snapshot_test_config.genesis_config_info.mint_keypair;
         let genesis_config = &snapshot_test_config.genesis_config_info.genesis_config;
+        let snapshot_version = bank_forks
```
Comment (ryoqun) [inline core/tests/bank_forks.rs:272]: Like [this](https://github.com/solana-labs/solana/pull/10536/files#r442013709), why is this needed?

File: ledger/src/snapshot_utils.rs [hunk 8]
```diff
@@ -37,14 +37,20 @@ pub const TAR_VERSION_FILE: &str = "version";
 const MAX_SNAPSHOT_DATA_FILE_SIZE: u64 = 32 * 1024 * 1024 * 1024; // 32 GiB
 const VERSION_STRING_V1_1_0: &str = "1.1.0";
 const VERSION_STRING_V1_2_0: &str = "1.2.0";
-const OUTPUT_SNAPSHOT_VERSION: SnapshotVersion = SnapshotVersion::V1_2_0;
+const DEFAULT_SNAPSHOT_VERSION: SnapshotVersion = SnapshotVersion::V1_2_0;
```
Comment (ryoqun) [inline ledger/src/snapshot_utils.rs:40]: nits: I think this `const` definition can now go away now. It's only used by `impl Default` for `SnapshotVersion`. Maybe just fold this into there? Named as `DEFAULT_...` here, this reads a bit redundant.

File: core/tests/bank_forks.rs (hunk 7 above)
Comment (svenski123) [reply core/tests/bank_forks.rs]: It's not, changes were solely to parameterize snapshot version.
I've tidied up the bank forks tests.

File: ledger/src/snapshot_utils.rs (hunk 8 above)
Comment (svenski123) [reply ledger/src/snapshot_utils.rs:40]: This is supposed to indicate the default output snapshot version - as new formats are added, they won't initially be made default until later - having everything together at the top should hopefully make this more apparent and easier to maintain.

Happy to remove it if you still feel it should go.

**Comment by svenski123** [comment]
Rebased and squashed

**Comment by ryoqun** [comment]
LGTM! Thanks for cleaning up!

File: core/tests/bank_forks.rs [hunk 9]
```diff
@@ -270,7 +312,7 @@ mod tests {
             if slot == saved_slot as u64 {
                 let options = CopyOptions::new();
                 fs_extra::dir::copy(accounts_dir, &saved_accounts_dir, &options).unwrap();
-                let snapshot_paths: Vec<_> = fs::read_dir(&snapshot_config.snapshot_path)
+                let snapshot_paths = fs::read_dir(snapshot_path)
```
Comment (ryoqun) [inline core/tests/bank_forks.rs:315]: nits: (I merged this because `nits` doesn't worth to block this useful pr. Feel free to create mini follow-up pr if this bothers you enough.) this started to be called `.last()` before assignment, so rename `snapshot_paths` => `snapshot_path`

File: core/tests/bank_forks.rs [hunk 10]
```diff
@@ -302,11 +340,12 @@ mod tests {
         // Purge all the outdated snapshots, including the ones needed to generate the package
         // currently sitting in the channel
         bank_forks.purge_old_snapshots();
-        let mut snapshot_paths = snapshot_utils::get_snapshot_paths(&snapshots_dir);
-        snapshot_paths.sort();
         assert_eq!(
```
Comment (ryoqun) [inline core/tests/bank_forks.rs:343]: nits: (I merged this because `nits` doesn't worth to block this useful pr. Feel free to create mini follow-up pr if this bothers you enough.) you can write `assert_eq!(..., true);` with `assert!(...);`

File: core/tests/bank_forks.rs [hunk 11]
```diff
@@ -386,21 +425,23 @@ mod tests {
...
+            let slots_to_snapshot = snapshot_test_config
                 .bank_forks
                 .get(snapshot_test_config.bank_forks.root())
-                .unwrap();
-            let slots_to_snapshot = rooted_bank.src.roots();
-            assert_eq!(slots_to_snapshot, expected_slots_to_snapshot);
+                .unwrap()
+                .src
+                .roots();
+            assert_eq!(
```
Comment (ryoqun) [inline core/tests/bank_forks.rs:437]: nits: (I merged this because `nits` doesn't worth to block this useful pr. Feel free to create mini follow-up pr if this bothers you enough.) you can write `assert_eq!(..., true);` with `assert!(...);`

File: core/tests/bank_forks.rs [hunk 12]
```diff
@@ -302,11 +340,12 @@ mod tests {
         // Purge all the outdated snapshots, including the ones needed to generate the package
         // currently sitting in the channel
         bank_forks.purge_old_snapshots();
-        let mut snapshot_paths = snapshot_utils::get_snapshot_paths(&snapshots_dir);
-        snapshot_paths.sort();
```
Comment (ryoqun) [inline core/tests/bank_forks.rs:306]: Nice. Yeah, we don't need to `.sort()` indeed. It's done inside `get_snapshot_paths()` already.

File: core/tests/bank_forks.rs (hunk 11 above)
Comment (svenski123) [reply core/tests/bank_forks.rs:437]: Fixed in #10721

File: core/tests/bank_forks.rs (hunk 10 above)
Comment (svenski123) [reply core/tests/bank_forks.rs:343]: Fixed in #10721

File: core/tests/bank_forks.rs (hunk 9 above)
Comment (svenski123) [reply core/tests/bank_forks.rs:315]: Fixed in #10721
